File: Data Structures & Algorithms/eating-bananas/submission-6.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int l = 0, r = Arrays.stream(piles).max().getAsInt();
        int res = r;

        while (l<=r) {
            int mid = l + (r - l)/2;
            int hours = 0;
            for (int pile: piles) {
                hours += Math.ceil((double) pile/mid);
            }

            if (hours > h) {
                l = mid + 1;
            } else {
                // res = Math.min(res, mid);
                res = mid; // the reason why this works is because every time this else condition is met, it means that it is a valid answer and also lowers the right bound, such that everytime this else condition the mid get smaller compared to previous iterates, meaing Math.min(res,mid) is extra and unesscary.
                r = mid - 1;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-6.py
class Solution:
    def findMin(self, nums: List[int]) -> int:
        # left = 0
        # right = len(nums) - 1
        # res =nums[0]
        # while (left < right):
        #     mid = left + (right - left)//2
        #     res = min(nums[left], nums[right], nums[mid])
        #     if (left == mid or right == mid):
        #         break
        #     if nums[left] > nums[right]:
        #         left = mid
        #     else:
        #         right = mid 

        # return res
        return min(nums)